tests/mock-network.ts

```diff
@@ -0,0 +1,235 @@
+    }
+  | {
+      error: unknown;
+    }
+  | {
+      implementation: () => void;
+    }
+);
+
+/**
+ * Type guard for determining whether the given value is an error object with a
+ * `message` property, such as an instance of Error.
+ *
+ * Copied from @metamask/utils.
+ *
+ * @param error - The object to check.
+ * @returns True or false, depending on the result.
+ */
+function isErrorWithMessage(error: unknown): error is { message: string } {
+  return typeof error === 'object' && error !== null && 'message' in error;
+}
+
+/**
+ * Get the error message from an unknown error object. If the error object has
+ * a `message` property, that property is returned. Otherwise, the stringified
+ * error object is returned.
+ *
+ * Copied from @metamask/utils.
+ *
+ * @param error - The error object to get the message from.
+ * @returns The error message.
+ */
+function getErrorMessage(error: unknown): string {
+  return isErrorWithMessage(error) ? error.message : String(error);
+}
+
+/**
+ * MockedNetwork provides a more streamlined interface around `nock` to mock
+ * JSON-RPC requests that are sent to a network.
+ */
+class MockedNetwork {
+  #networkClientConfiguration: NetworkClientConfiguration;
+
+  #requestMocks: JsonRpcRequestMock[];
+
+  #nockScope: nock.Scope;
+
+  /**
+   * Makes a new MockedNetwork.
+   *
+   * @param args - The arguments.
+   * @param args.networkClientConfiguration - Details about the network which
+   * tell this class which URL to mock.
+   * @param args.mocks - Objects which specify the requests to mock and the
+   * responses to use for those requests.
+   */
+  constructor({
+    networkClientConfiguration,
+    mocks = [],
+  }: {
+    networkClientConfiguration: NetworkClientConfiguration;
+    mocks: JsonRpcRequestMock[];
+  }) {
+    this.#networkClientConfiguration = networkClientConfiguration;
+    this.#requestMocks = mocks;
+    const rpcUrl =
+      networkClientConfiguration.type === 'infura'
+        ? `https://${networkClientConfiguration.network}.infura.io`
+        : networkClientConfiguration.rpcUrl;
+    this.#nockScope = nock(rpcUrl);
+  }
+
+  /**
+   * Mocks all of the requests that have been specified via the constructor.
+   */
+  enable() {
+    for (const requestMock of this.#requestMocks) {
+      this.#mockRpcCall(requestMock);
+    }
+  }
+
+  /**
+   * Mocks a JSON-RPC request sent to the network.
+   *
+   * @param requestMock - Specifies the request to mock and how that request
+   * should respond.
+   * @returns The resulting Nock scope.
+   */
+  #mockRpcCall(requestMock: JsonRpcRequestMock): nock.Scope {
+    // eth-query always passes `params`, so even if we don't supply this
+    // property, for consistency with makeRpcCall, assume that the `body`
+    // contains it
+    const { method, params = [], ...rest } = requestMock.request;
+    const httpStatus =
+      'response' in requestMock && 'httpStatus' in requestMock.response
```

Comment:
Curious why this condition was updated (I see that the old version allowed setting a status without a response)

Reply:
Yeah, good question. 

I updated this because when I wanted to use MockedNetwork for other tests, I noticed I couldn't specify just an `error` property in a request mock. It seems that in borrowing the [interface from FakeProvider](https://github.com/MetaMask/core/blob/ea879d268a8862f29cd1227fe8d7236998e3a048/packages/network-controller/tests/fake-provider.ts#L44) I mistakenly made `response` required. So I've fixed that here so that it is possible to specify only `error` if you want to (in fact, `response` and `error` are mutually exclusive).

In terms of whether `response` should be necessary to specify `httpStatus`, the way that the `error` property works here, which is different from FakeProvider, is that it will throw an error rather than reject a promise with an error. So if you specify `error`, `httpStatus` has no effect, because the response never happens. So that's why `error` and `response` are mutually exclusive.

Does that make sense?

Reply:
OK, I think I follow. Though I'm not sure how this meshes with the `implementation` option, which seemingly might take the place of both a response and an error.

Perhaps this could be moved into the `else` block below on line 169? Then we would know from context that there is no error, so it should be safe to assume that a status is allowed.

Reply:
Ah. I just checked again and I'm not even using the `implementation` property. So I've removed that here: [cd21290](https://github.com/MetaMask/core/pull/1508/commits/cd212909dd639040f8fc108b08e56ca4ae494c18). And I've moved `httpStatus` down here: [f5114a3](https://github.com/MetaMask/core/pull/1508/commits/f5114a38eb8acc6d05710ecbf1fe73ec85389748)